.github/workflows/manual-sync.yml

```diff
@@ -0,0 +1,78 @@
+name: Sync from osu! upstream
+
+# Workflow dispatch can take inputs like any other github.event
+# As part of #5090 - we should evaluate adding inputs and outputs to refine this.
+on:
+  workflow_dispatch:
+    inputs:
+      force:
+        description: "Overwrite any changes in the target repository? (true/false)"
+        required: true
+        default: "false"
+
+jobs:
+  sync:
+    name: Execute Sync Task
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Configure Git to use your name
+        shell: bash
+        run: |
+          set -eo pipefail
+
+          #FIXME: We're using the noreply email of GitHub for now. Let's figure out a way to get the committer's email.
+          git config --global user.name ${{ github.actor }}
+          git config --global user.email ${{ github.actor }}@users.noreply.github.com
+
+
+      - name: Begin Pull Task
+        shell: bash
+        run: |
+         set -eo pipefail
+
+         # sanitize github input into boolean
+         # https://stackoverflow.com/questions/14700579/bash-converting-string-to-boolean-variable
+         function boolean () {
+            case $1 in
+              true) echo true ;;
+              false) echo false ;;
+              *) echo "Unknown operand \"$1\"" 1>&2; exit 1 ;;
+            esac
+         }
+
+         #Upstream repository URL, this is where everything ties down.
+         UPSTREAM_REPO="https://github.com/ppy/osu-wiki"
+         BACKUP_BRANCH_NAME="backup-$(git branch --show-current)"
+         BRANCH_NAME="$(git branch --show-current)"
+
+         git remote add upstream $UPSTREAM_REPO
+         git fetch upstream
+
+         if [ ! $(boolean ${{ github.event.inputs.force }}) ]; then
+           # Make a backup then rebase all changes
+           git branch $BACKUP_BRANCH_NAME
+           git reset --hard upstream/master
+           git rebase upstream/master
+           git merge --squash $BACKUP_BRANCH_NAME -m "Merge changes from Upstream to $(BRANCH_NAME)"
+
+           git push origin +$BRANCH_NAME
```

Comment:
I'm reading this again and I'm thoroughly confused.

* What's the rebase doing there? `git reset --hard` moves the branch pointer to the given ref, does it not? There *is* nothing to rebase onto at that point...?
* `git merge --squash` [is documented to not merge](https://git-scm.com/docs/git-merge#Documentation/git-merge.txt---squash), so I'm not sure what the push is supposed to achieve?

I've tried these commands locally and it doesn't look like this does what it claims to be. How does this work?

Reply:
I've tried out some of these commands locally so I was a bit sure how they'll work. Looks like I have to redo this again, thanks for the catch.

Reply:
The idea behind these is 

- we make a backup branch to keep our changes
- we reset this tree based on `upstream/master`
- Call a rebase to verify if we do have the current HEAD set to `upstream/master`. If not, rebase will rebase all the changes in.
- merge the changes in, squashed (though I'll just make this a merge commit now in the new iteration).

Reply:
I'm still lost. What is the goal of this tool? Is the goal to sync the master branch, or is it to help users update their "feature" branches with changes in master?

If it's the first one, this should be no more than:

```
git checkout master // to make sure, i guess
git branch backup
git pull --ff-only upstream master // refuse to update if user has new commits that never were in upstream/master
```

Alternatively, depending on what we want, that last one should be `git reset --hard upstream/master`, if the goal is to just bring the user back to a good state and not care that they had their own stuff merged into master. After all they have a backup branch.

If it's the second one, this should be no more than:

```
git merge upstream/master
```

Reply:
> I'm still lost. What is the goal of this tool? Is the goal to sync the master branch, or is it to help users update their "feature" branches with changes in master?
> 
> If it's the first one, this should be no more than:
> 
> ```
> git checkout master // to make sure, i guess
> git branch backup
> git pull --ff-only upstream master // refuse to update if user has new commits that never were in upstream/master
> ```
> 
> Alternatively, depending on what we want, that last one should be `git reset --hard upstream/master`, if the goal is to just bring the user back to a good state and not care that they had their own stuff merged into master. After all they have a backup branch.
> 
> If it's the second one, this should be no more than:
> 
> ```
> git merge upstream/master
> ```

it's designed to sync their changes regardless of what branch they are in, so we want to avoid relying too much on `master` if such a case we have some web editor people using master as their working tree.

The reason I want the branch in a clean state for the non-destructive flow is to make sure the changes are integrated "safely", so I did a reset first then merge the changes in from the backup branch. Though i can see the `no-ff` part more feasible but I fear it may break easily and leave our tree in a very undesirable state.

Reply:
@bdach I recommend you suggest a change for this part that works much more better, but as far as it goes I think this is the safest I can go :D

Reply:
I'm not sure what more to suggest beyond what I already have. I'm still lost on why the current code is doing a fetch, then a hard reset and finally a pull. The pull is entirely redundant (pull = fetch + merge; the fetch has already happened manually, and the merge does nothing as the branch pointer was already moved to that same ref being merged, so I fail to see why it's "safer").

Reply:
IIRC during my testing inside GitHub Codespaces Git doesn't pull the changes from `upstream/master`'s HEAD, though I could be wrong, I'll probably refactor this to use your suggestions and see if it'll bite us in the long run

Reply:
1eb3fa4 should have this refactored now, thanks for your insight, please RFC again if you're not satisifed @bdach :D